[EventTrace] AddEventPipe filter data parsing

mdh1418 opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/102572

This PR follows-up on https://github.com/dotnet/runtime/pull/102612#discussion_r1649499591 and https://github.com/dotnet/runtime/pull/102612#discussion_r1651605026 by extending the FilterData parsing to cover EventPipe scenarios.

## Testing

Checked the value of  [l64ClientSequenceNumber](https://github.com/dotnet/runtime/blob/6d23ef4d68bbcdb38fdc22218d1073c5083ac6a1/src/coreclr/vm/eventtrace.cpp#L2423) after enabling an EventPipe session with ` dotnet trace collect -n sampleapp --providers "Microsoft-Windows-DotNETRuntime:800000:5:mihw=whim;GCSeqNumber=2024"` and saw `0x7e8`.

mdh1418 committed:
[EventTrace] Add EventPipe FilterData parsing

mdh1418 committed:
Add NativeAOT counterpart

noahfalk commented on src/coreclr/vm/eventtrace.cpp:
```diff
@@ -2328,6 +2328,72 @@ enum CallbackProviderIndex
 };
 
+// EventFilterType identifies the filter type used by the PEVENT_FILTER_DESCRIPTOR
+enum EventFilterType
```
Rather than including the ETW enum I'd focus on values we will accept and what we expect those enumerants to mean. For example:
```suggestion
enum EventFilterType
{
    // data should be pairs of UTF8 null terminated strings all concatenated together.
    // The first element of the pair is the key and the 2nd is the value. We expect one of the
    // keys to be the string "GCSeqNumber" and the value to be a number encoded as text.
    // This is the standard way EventPipe encodes filter values
    StringKeyValueEncoding = 0
    // data should be an 8 byte binary LONGLONG value
    // this is the historic encoding defined by .NET Framework for use with ETW
    LongBinaryClientSequenceNumber = 1
}
```

mdh1418 replied:
Changed! Thanks

noahfalk commented on src/coreclr/vm/eventtrace.cpp:
```diff
@@ -2328,6 +2328,72 @@ enum CallbackProviderIndex
+
+            const char* value = buffer;
+            buffer += strlen(value) + 1;
+
```
To be safe against a mal-formed buffer we should do a bounds check on the value portion too:
```suggestion
if(buffer > buffer_end)
    break;
```

noahfalk commented on src/coreclr/nativeaot/Runtime/eventtrace.cpp:
```diff
@@ -94,6 +94,42 @@ enum CallbackProviderIndex
+    if (FilterData->Type == ClientSequenceNumber && FilterData->Size == sizeof(LONGLONG))
+    {
+        *pClientSequenceNumber = *(LONGLONG *) (FilterData->Ptr);
+    }
```
NativeAOT supports EventPipe too so unless there is particular reason not to I'd support the same parsing options that we have on CoreCLR.

mdh1418 replied:
I left it out since there was a comment that `// NativeAOT currently only supports forcing a GC with ManagedHeapCollectKeyword via ETW`, so this path won't get hit until forcing a GC via EventPipe is supported on NativeAOT. I figured the parsing could be added once/if the EventPipe scenario is supported.

I don't mind adding the parsing, if that is preferred. @LakshanF @elinor-fung do y'all happen to know if NativeAOT will ever support forcing a GC with the keyword via EventPipe?

noahfalk replied:
I'm not sure why that limitation is there and I'd be pretty surprised if its fundamental. My guess was that it was omitted just for lack of time to implement/test something.

brianrob replied:
Agreed with @noahfalk here.  I suspect this was an exercise in scoping more than anything else.

mdh1418 replied:
Thanks! Added the EventPipe parsing to the nativeaot counterpart.

noahfalk reviewed (approved):
Some suggestions inline, but the approach seems good.

mdh1418 committed:
Address feedback

Only keep implemented filter types in enum
Add EventPipe parsing to nativeaot counterpart
Add bounds check on value portion of buffer

mdh1418 committed:
Guard against non null-terminated buffers

mdh1418 commented on src/native/eventpipe/ep-provider.c:
```diff
@@ -421,7 +421,7 @@ provider_invoke_callback (EventPipeProviderCallbackData *provider_callback_data)
 			buffer_size = j + 1;
 
-		ep_event_filter_desc_init (&event_filter_desc, (uint64_t)buffer, buffer_size, 0);
+		ep_event_filter_desc_init (&event_filter_desc, (uint64_t)buffer, buffer_size, /* EventFilterType.EventPipePayload */ 2);
```
Done, thanks!

mdh1418 committed:
Fix macro guard

noahfalk commented on src/coreclr/nativeaot/Runtime/eventtrace.cpp:
```diff
@@ -94,6 +94,67 @@ enum CallbackProviderIndex
+            size_t value_len = strnlen(value, buffer_end - buffer);
+            buffer += value_len + 1;
+
+            if (buffer >= buffer_end)
```
I think this should be '>' no? Maybe I have the wrong mental model but this is what I was expecting:
FilterData->Ptr = 0x1000000, pointing to the UTF8 string "GCSeqNumber\025\0"
FilterData->Size = 15
key = 0x1000000
key_len = 11
value = 0x100000c (+12)
value_len = 2
buffer = 0x100000f (+15)
buffer_end = 0x100000f (+15)

mdh1418 replied:
Ah, you're right, I tricked myself thinking that the break conditions would be the same.

noahfalk reviewed (approved):
I think you may have an off-by-one error, but looks good otherwise 👍

mdh1418 committed:
Fix break condition
